src/conversation_engine/engine.py

```diff
@@ -125,29 +149,40 @@ def aborting(self, timeout, timeout_callback):
                     timeout_callback,
                     oneshot=True)
 
-    def initialize_semantics(self, semantics):
-        self._current_semantics = semantics
-        rospy.loginfo("Initialized semantics: {}".format(self._current_semantics))
-
-    def update_semantics(self, semantics, missing_field_path):
+    @staticmethod
+    def update_semantics(old_semantics, semantics, missing_field_path):
+        """
+        Update the dictionary in old_semantics at the path specified in missing_field_path with the content of semantics
+        :param old_semantics: Original dictionary
+        :type old_semantics: Dict[str, str]
+        :param semantics: Data to be entered into old_semantics at missing_field_path
+        :type semantics: Dict[str, str]
+        :param missing_field_path: Where to add semantics into old_semantics
+        :type missing_field_path: str
+        :return: a new dictionary that merges the old and new semantics
+        >>> old = {'actions': [{'action': 'find', 'object': {'type': 'person', 'id': 'lars'}}]}
+        >>> update = {'id': 'couch'}
+        >>> ConversationState.update_semantics(old, update, 'actions[0].source-location')
+        {'actions': [{'action': 'find', 'source-location': {'id': 'couch'}, 'object': {'type': 'person', 'id': 'lars'}}]}
+        """
         # I assume semantics is the exact information requested and supposed to go in place of the field indicated by
         # the missing information path
 
+        to_be_updated = old_semantics.copy()
```

Comment:
You use a copy and not a deepcopy. Furtherone, you update fields in the copy of the dict. Does this affect the original? Is this the intended behavior?

Not entirely part of this merge request but: I find lines 174 - 185 very hard to understand. There seem to be quite some assumptions on the contents of the dicts. It would be useful to make this clear in some comments.